nginx/src/charmlibs/nginx/config.py

```diff
@@ -0,0 +1,627 @@
+    """Custom headers to include in the proxied request."""
+    modifier: NginxLocationModifier = ''
+    """The Nginx location modifier."""
+    is_grpc: bool = False
+    """Whether to use gRPC proxying (i.e. `grpc_pass` instead of `proxy_pass`)."""
+    upstream_tls: bool | None = None
+    """Whether to connect to the upstream over TLS (e.g., https:// or grpcs://)
+    If None, it will inherit the TLS setting from the server block that the location is part of.
+    """
+
+
+@dataclass
+class NginxUpstream:
+    """Represents metadata needed to construct an Nginx `upstream` block."""
+
+    name: str
+    """Name of the upstream block."""
+    port: int
+    """Port number that all backend servers in this upstream listen on.
+
+    Our coordinators assume that all servers under an upstream share the same port.
+    """
+    group: str | None = None
+    """Group that this upstream belongs to.
+
+    Used for mapping multiple upstreams to a single group of backends (loadbalancing between all).
+    If you leave it None, this upstream will be routed to all available backends
+    (loadbalancing between them).
+    """
+
+
+class NginxConfig:
+    """Responsible for building a Nginx configuration file."""
+
+    _pid = '/tmp/nginx.pid'  # noqa
+
+    def __init__(
+        self,
+        server_name: str,
+        upstream_configs: list[NginxUpstream],
+        server_ports_to_locations: dict[int, list[NginxLocationConfig]],
+        enable_health_check: bool = False,
+        enable_status_page: bool = False,
+        supported_tls_versions: list[str] | None = None,
+        ssl_ciphers: list[str] | None = None,
+        worker_processes: int = 5,
+        worker_connections: int = 4096,
+        proxy_read_timeout: int = 300,
+        proxy_connect_timeout: str = '5s',
+    ):
+        """Constructor for a Nginx config generator object.
+
+        Args:
+            server_name: The name of the server (e.g. fqdn), which is used to identify
+              the server in Nginx configurations.
+            upstream_configs: List of Nginx upstream metadata configurations used to generate Nginx
+              `upstream` blocks.
+            server_ports_to_locations: Mapping from server ports to a list of Nginx location
+              configurations.
+            enable_health_check: If True, adds a `/` location that returns a basic 200 OK response.
+            enable_status_page: If True, adds a `/status` location that enables `stub_status` for
+              basic Nginx metrics.
+            supported_tls_versions: list of supported tls protocol versions.
+            ssl_ciphers: ssl ciphers.
+            worker_processes: Number of nginx worker processes to spawn.
+            worker_connections: Max number of worker connections
+            proxy_read_timeout: Proxy read timeout.
+            proxy_connect_timeout: Proxy connect timeout.
+
+        Example:
+            .. code-block:: python
```

Comment:
```suggestion
        Example::
```
I think a trailing `::` does the trick here. The indentation below needs to be fixed too tho. Not currently a docs build error, since we don't include `__init__` docstrings, only the class docstring -- so maybe move this docstring to the class?